common/utilities/tests/test_logger.py

```diff
@@ -0,0 +1,124 @@
+        log.load_global_log_config()
+        log.IntegrationAdaptersLogger()._format_and_write(
+            message="{yes} {no} {maybe}",
+            values={'yes': 'one', 'no': 'two', 'maybe': 'three'},
+            request_id=10,
+            correlation_id=5,
+            level=logging.INFO,
+            process_key_num="100"
+        )
+
+        output = mock_std.getvalue()
+
+        # Check that each value has spaces
+        self.assertIn(' CorrelationId=5 ', output)
+        self.assertIn(' LogLevel=INFO ', output)
+        self.assertIn(' ProcessKey=SYS100', output)
+        self.assertIn(' RequestId=10 ', output)
+        self.assertIn(' maybe=three ', output)
+        self.assertIn(' yes=one ', output)
+        self.assertIn(' pid=', output)
+
+    @patch('sys.stdout', new_callable=io.StringIO)
+    def test_format_and_write_empty_vals(self, mock_std):
+        log.load_global_log_config()
+        log.IntegrationAdaptersLogger()._format_and_write(
+            message="{yes} {no} {maybe}",
+            values={'yes': 'one', 'no': 'two', 'maybe': 'three'},
+            request_id=None,
+            correlation_id=None,
+            level=logging.INFO,
+            process_key_num="100"
+        )
+
+        output = mock_std.getvalue()
+        self.assertIn(' LogLevel=INFO ', output)
+        self.assertIn(' ProcessKey=SYS100', output)
+        self.assertIn(' maybe=three ', output)
+        self.assertIn(' pid=', output)
+        self.assertIn(' yes=one ', output)
+
+    def test_log_levels(self):
+        logger = log.IntegrationAdaptersLogger()
+        logger._format_and_write = MagicMock()
+        with self.subTest("INFO"):
+            logger.info("{yes}", {'yes': 'no'}, "100", "REQ", 313)
+            logger._format_and_write.assert_called_with("{yes}", {'yes': 'no'}, "100", "REQ", 313, logging.INFO)
+
+        with self.subTest("AUDIT"):
+            logger.audit("{yes}", {'yes': 'no'}, "100", "REQ", 313)
+            logger._format_and_write.assert_called_with("{yes}", {'yes': 'no'}, "100", "REQ", 313, log.AUDIT)
+
+        with self.subTest("WARNING"):
+            logger.warning("{yes}", {'yes': 'no'}, "100", "REQ", 313)
+            logger._format_and_write.assert_called_with("{yes}", {'yes': 'no'}, "100", "REQ", 313, logging.WARNING)
+
+        with self.subTest("ERROR"):
+            logger.error("{yes}", {'yes': 'no'}, "100", "REQ", 313)
+            logger._format_and_write.assert_called_with("{yes}", {'yes': 'no'}, "100", "REQ", 313, logging.ERROR)
+
+        with self.subTest("CRITICAL"):
+            logger.critical("{yes}", {'yes': 'no'}, "100", "REQ", 313)
+            logger._format_and_write.assert_called_with("{yes}", {'yes': 'no'}, "100", "REQ", 313, logging.CRITICAL)
+
+    @patch('sys.stdout', new_callable=io.StringIO)
+    def test_empty_values(self, mock_stdout):
+        log.load_global_log_config()
+        logger = log.IntegrationAdaptersLogger()
+        with self.subTest("Empty info log"):
```

Comment:
Optional: for this and the previous test, you could do some sort of parametrising-y thing to refactor out the common parts in each test. Though this probably/possibly would make the test less readable, so not sure if it's worth it.

Reply:
Yeah would definitely save lines, though I'll keep it as it is if you think it might make it harder to read